contributors/design-proposals/sig-storage/volume-topology-scheduling.md

```diff
@@ -0,0 +1,580 @@
+fast data access, and another PVC can point to a local HDD for logging.  Since PVC
+binding happens without considering if multiple PVCs are related, it is very likely
+for the two PVCs to be bound to local disks on different nodes, making the pod
+unschedulable.
+* For multizone clusters and deployments requesting multiple dynamically provisioned
+zonal PVs, each PVC Is provisioned independently, and is likely to provision each PV
+In different zones, making the pod unschedulable.
+
+To solve the issue of initial volume binding and provisioning causing an impossible
+pod placement, volume binding and provisioning should be more tightly coupled with
+pod scheduling.
+
+
+## Background
+In 1.7, we added alpha support for local PVs (TODO: link) with node affinity.
+You can specify a PV object with node affinity, and if a pod is using such a PV,
+the scheduler will evaluate the PV node affinity in addition to the other
+scheduling predicates.  So far, the PV node affinity only influences pod
+scheduling once the PVC is already bound.  The initial PVC binding decision was
+unchanged.  This proposal addresses the initial PVC binding decision.
+
+
+## Design
+The design can be broken up into a few areas:
+* User-facing API to invoke new behavior
+* Integrating PV binding with pod scheduling
+* Binding multiple PVCs as a single transaction
+* Recovery from kubelet rejection of pod
+* Making dynamic provisioning topology-aware
+
+For the alpha phase, only the user-facing API and PV binding and scheduler
+integration are necessary.  The remaining areas can be handled in beta and GA
+phases.
+
+### User-facing API
+This new binding behavior will apply to all unbound PVCs and can be controlled
+through install-time configuration passed to the scheduler and controller-manager.
+
+In alpha, this configuration can be controlled by a feature gate,
+`VolumeTopologyScheduling`.
+
+#### Alternative
+An alternative approach is to only trigger the new behavior only for volumes that
+have topology constraints.  A new annotation can be added to the StorageClass to
+indicate that its volumes have constraints and will need to use the new
+topology-aware scheduling logic.
+
+```
+// Value is “true” or “false”.  Default is false.
+const AlphaVolumeTopologySchedulingAnnotation = “volume.alpha.kubernetes.io/topology-scheduling”
+```
+
+While this approach can let us introduce the new behavior gradually, it has a
+few downsides:
+* StorageClass will be required to use this new logic, even if dynamic
+provisioning is not used (in the case of local storage).
+* We have to maintain two different paths for volume binding.
+* We will be depending on the storage admin to correctly configure the
+StorageClasses.
+
+For those reasons, we prefer to switch the behavior for all unbound PVCs and
+clearly communicate these changes to the community before changing the behavior
+by default in GA.
+
+### Integrating binding with scheduling
```

Comment:
From software engineering point of view, mixing the responsibilities of modules and removing the isolation of responsibilities is not a great practice. With this design, the scheduler process would perform volume binding which is not one of the usual responsibilities of a cluster scheduler. Other than removing the modularity, this could introduce runtime problems as well. For example, a null dereference in volume binding logic will cause a scheduler crash. High memory or CPU consumption in volume binding logic could cause scheduler to run out of memory or starve.

Reply:
Let me clarify that I am not saying that we shouldn't move forward with this design, but when we consider pros and cons of various approaches, this should be considered as a negative point of this design.

Reply:
I think we do need a tighter coupling (somehow) between scheduler and volume binding in order to address the problems I outlined in the beginning.  At a high level:
* Scheduler needs to be aware of a volume's topology wrt the node when it is scheduling a pod
* The volume binding decision needs to respect the pod's other scheduling constraints

This design proposes a very tight coupling by putting the volume binding directly into the scheduler.  You are correct in that it's expanding the scheduler's responsibilities to more than just Pod->Node binding.  The major risk I see with this current approach is that it's trying to add in the volume binding using the existing scheduling mechanics, which may be designed with the assumption that Pod->Node binding is the only thing the scheduler does.  So this approach breaks when you consider the scenario where volumes are unbound, but the user has already manually scheduled the Pod to a Node.  Fixing that I think is going to require some fundamental changes in the scheduler, which could greatly increase the scope of this design.  But it may be something still worth considering, especially if more objects also need to integrate their binding with scheduling.  I know @vishh and @yguo0905 are also looking into extending the scheduler for binding other resources (although may not be as complex as volume binding), and they may also run into the same issues as me.

I did also consider an alternate approach with loose coupling, where it tries to keep scheduler and PV controller separate.  It's similar to this [alternative discussion](https://github.com/kubernetes/community/pull/1054/files#diff-fb55f38843406f82a21ecd36f75efd72R456).  The basic idea is the PV controller waits for the Pod.Spec.NodeName to be set before it does the binding.  However, even though the binding process is decoupled from the scheduler, the binding decision is much more prone to split brain problems.  We still need a new scheduler predicate so that the scheduler can pick an appropriate node, but because the binding happens completely asynchronously in the PV controller, it's possible that the actual binding will pick a different volume, or fail completely.  Then you need a way to tell the scheduler to revert its PV cache update, and also a way for the PV controller to tell kubelet to reject the Pod so that it can be sent back through scheduling.  So while this approach keeps the binding process and code separate, I found that it:
* is more prone to split brain failures due to scheduler and PV controller trying to make the same decisions, but operating on different and out-of-sync caches.
* required very tight, complex, and delicate coordination between scheduler, PV controller, and kubelet processes to handle recovering from these split brain failures.

I am open to revisiting this alternative if you think it looks more promising than the current approach, but from my initial investigation, it seemed much more complex and fragile.

Reply:
> With this design, the scheduler process would perform volume binding which is not one of the usual responsibilities of a cluster scheduler.

I am not sure what "cluster scheduler" is, however with this PR we're moving from Pod scheduler to Pod and PV scheduler. And I am sure more resources will come in future.

E.g. setting `pod.spec.nodeName` should not bypass the scheduler as it should schedule the storage too.

Reply:
I think we could benefit from extending the scheduler to support multiple object scheduling/binding in a modular manner.  Each module would need to:
* Given a Pod object, evaluate if it needs to run through the scheduler
* Register assume/bind functions
* Register rollback function

Of course, there is a ton of detail that needs to be worked out